textcursor.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gxui
+
+import "github.com/google/gxui/interval"
+
+// TextCursor represents a text position and selection. A selection is defined as
+// the pair of cursor indices sorted least to greatest. When transforming a
+// cursor with no selection, mark should equal index.
+type TextCursor struct {
```

Comment:
Cursor.Index is the location of the caret. Given that the words cursor and caret refer to the same idea, the rationale for the field name Index is that a single cursor/caret can only exist at a single position.

I am somewhat opposed to the names From and To as it conflates the idea of a cursor/caret and a selection. This is also why I omitted the Start and End methods. A selection is retrieved via TextCursor.Selection which is currently simple enough to be a pair of ordered indices, but still distinct from the cursor itself. Would TextCaret.Index be more obvious?

Regarding Cursor.Mark, I'm not totally convinced on that name either. Mark is meant to represent storing a cursor position but is still volatile. During transforms, mark is never meant to be operated on independently. All transforms should either adjust only the cursor index, or apply the same change to cursor index and cursor mark. I've been considering how to enforce this contract through what methods are available on TextCursor as well.

If you look at the TextCursorSlice.Transform, this uniformly transforms index and mark, no special cases. This is essentially what TextCursor.Offset does as well, allowing the cursor, and any selection implied, to move around.

If you also look at TransformCarets (which I forgot to rename to TransformSelection, will follow up with that change), it only operates on the cursor index.

So those transform methods operate within the bounds of the contract I describe, but the TextCursor itself does not yet.

What I've considered next to assure TextCursor makes this clear is to provide the following constructors instead in a follow up commit (since it will touch more code elsewhere). `NewTextCursor(index int) Cursor` and `NewTextCursorSelection(start, end int)`.

So the distinction I'm trying to make here is that a TextCursor is not a half-closed interval up until the point you call `TextCursor.Selection()`. It's just happen-stance they are stored side by side in the same struct due to a selection being trivially represented by an additional index.

I think this is an important distinction, particularly when you look at all the method on textboxcontroller working with selections and carets and all the ints going around everywhere. A new caret creates a hidden selection that could be retrieved with a selection method and other weirdness.

I'm also against the redundancy of Cursor.Caret, as I hope I clarified what a TextCursor is above but perhaps renaming to TextCaret would make things totally obvious.

> I'm not sure it's worth bringing in the concept of selection into the TextCursor documentation 

I agree. That's more or less there due to maintaining a similar constructor to the old code.

I love the ascii art :)